Add support for strategy-based UUID auto-generation (@DynamoDbAutoGeneratedUuid)

anasatirbasa opened the pull request:
## Description

Added support for **strategy-based UUID auto-generation** on `@DynamoDbAutoGeneratedUuid`, aligned with legacy V1 semantics from `@DynamoDBGeneratedUuid(DynamoDBAutoGenerateStrategy)`.

The V2 annotation now supports:

- `strategy = ALWAYS` (default): generate a new UUID on every write
- `strategy = CREATE`: generate only when the attribute is missing/empty

This provides the V1 "create-only" behavior without introducing a separate key annotation or extension.

### Important Rules

- `@DynamoDbAutoGeneratedUuid` is valid only for `String` attributes.
- `strategy = CREATE` can be used on both key and non-key attributes.
- `strategy = ALWAYS` preserves existing V2 behavior (regenerate on write), so backward compatibility is maintained for existing users of `@DynamoDbAutoGeneratedUuid`.

When `strategy = CREATE` is used, generation happens only if the value is absent/blank; existing non-blank values are preserved.

### UpdateBehavior Notes

`@DynamoDbUpdateBehavior` still applies only to `UpdateItem` expression behavior (not `PutItem`).

- **Primary keys (PK/SK):**
  handled as key attributes in request construction; update behavior does not change primary-key immutability semantics.

- **Secondary index keys / non-key attributes:**
  update behavior (`WRITE_ALWAYS` / `WRITE_IF_NOT_EXISTS`) influences how generated values are written in update expressions.

---

## Motivation and Context

This PR remains related to issue [#5497](https://github.com/aws/aws-sdk-java-v2/issues/5497), but the implementation direction changed:

- Instead of introducing a new `@DynamoDbAutoGeneratedKey` API, this PR extends the existing `@DynamoDbAutoGeneratedUuid` with a strategy.
- This follows the V1 model more closely, where behavior is strategy-driven (`ALWAYS` vs `CREATE`).
- Existing users who rely on regenerate-on-write semantics remain unaffected because default strategy is `ALWAYS`.
- Users needing stable "generate only when missing" semantics can opt into `strategy = CREATE`.

This avoids API duplication and removes annotation-selection ambiguity.

---

## Modifications

- Added `DynamoDbAutoGenerateStrategy` enum with:
  - `ALWAYS`
  - `CREATE`
- Extended `@DynamoDbAutoGeneratedUuid` with a `strategy` parameter (default `ALWAYS`).
- Updated UUID generation flow in `AutoGeneratedUuidExtension` to apply per-attribute strategy at write time.
- Updated `AutoGeneratedUuidTag` to propagate annotation strategy into table metadata.
- Removed the separate AutoGeneratedKey-based implementation path (annotation/extension/tag/conflict model), consolidating behavior under the existing UUID annotation.
- Kept/updated type validation (`String` only) and strategy behavior validation paths.
- Updated docs and examples to use `@DynamoDbAutoGeneratedUuid(strategy = CREATE)` for create-only semantics.

---

## Testing

- Updated unit and functional tests to cover strategy-driven behavior end-to-end.
- Added/updated tests for:
  - `ALWAYS` behavior (regenerate on writes)
  - `CREATE` behavior (generate only when missing/empty; preserve existing values)
  - key and non-key usage scenarios
  - interaction with update flows and `@DynamoDbUpdateBehavior`
- Removed obsolete tests tied to the removed AutoGeneratedKey/conflict path.
- Test suite validates V1-style semantics in V2 while preserving backward compatibility for existing `@DynamoDbAutoGeneratedUuid` users.


#### Test Coverage on modified classes:
<img width="817" height="354" alt="image" src="https://github.com/user-attachments/assets/b8ce6ffb-6b6a-4355-a4a7-e25966c83774" />



## Test Coverage Checklist  

| Scenario | Done | Comments if Not Done |
|---------|:----:|---------------------|
| **1. Different TableSchema Creation Methods** | | |
| a. TableSchema.fromBean(Customer.class) | [x] | |
| b. TableSchema.fromImmutableClass(Customer.class) for immutable classes | [x] | |
| c. TableSchema.documentSchemaBuilder().build() | [ ] | |
| d. StaticTableSchema.builder(Customer.class) | [x] | |
| **2. Nesting of Different TableSchema Types** | | |
| a. @DynamoDbBean with annotated auto-generated key | [x] | |
| b. @DynamoDbImmutable with annotated auto-generated key | [x] | |
| c. Auto-generated key combined with partition/sort key | [x] | |
| **3. CRUD Operations** | | |
| a. scan() | [ ] | |
| b. query() | [x] | |
| c. updateItem() preserves existing value or generates when absent | [x] | |
| d. putItem() with no key set (auto-generation occurs) | [x] | |
| e. putItem() with key set manually | [x] | |
| f. getItem() retrieves auto-generated key | [x] | |
| g. deleteItem() | [ ] | |
| h. batchGetItem() | [ ] | |
| i. batchWriteItem() | [ ] | |
| j. transactGetItems() | [ ] | |
| k. transactWriteItems()  | [ ] | |
| **4. Data Types and Null Handling** | |  |
| a. Annotated attribute is null → key auto-generated | [x] | |
| b. Annotated attribute non-null → value preserved | [x] | |
| c. Validation rejects non-String annotated attribute | [x] | |
| **5. AsyncTable and SyncTable** | | |
| a. DynamoDbAsyncTable Testing | [x] | |
| b. DynamoDbTable Testing | [x] | |
| **6. New/Modification in Extensions** | | |
| a. Works with other extensions like VersionedRecordExtension | [x] | |
| b. Test with Default Values in Annotations | [ ] | |
| c. Combination of Annotation and Builder passes extension | [ ] | |
| **7. New/Modification in Converters** | | |
| a. Tables with Scenario in ScenarioSl No.1 (All table schemas are Must) | [ ] | |
| b. Test with Default Values in Annotations | [ ] | |
| c. Test All Scenarios from 1 to 5 | [ ] | |

## Types of changes  
- [ ] Bug fix (non-breaking change which fixes an issue)  
- [x] New feature (non-breaking change which adds functionality)  

## Checklist  
- [x] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document  
- [x] Local run of `mvn install` succeeds  
- [x] My code follows the code style of this project  
- [ ] My change requires a change to the Javadoc documentation  
- [ ] I have updated the Javadoc documentation accordingly  
- [x] I have added tests to cover my changes  
- [x] All new and existing tests passed  
- [x] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.  
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)  

## License  
- [x] I confirm that this pull request can be released under the Apache 2 license

marcusvoltolim reviewed (changes requested):
No tests with `@DynamoDbSortKey`

anasatirbasa commented on services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/enhanced/dynamodb/functionaltests/AutoGeneratedKeyRecordTest.java:
```diff
@@ -0,0 +1,535 @@
+            GsiPartitionKeyBean out = table.getItem(r -> r.key(k -> k.partitionValue("id123")));
+            assertValidUuid(out.getGsiPk());
+        } finally {
+            getDynamoDbClient().deleteTable(b -> b.tableName(tn));
```
Fixed.

anasatirbasa commented on services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/extensions/AutoGeneratedKeyExtension.java:
```diff
@@ -0,0 +1,184 @@
+    private void insertUuidIfMissing(Map<String, AttributeValue> itemToTransform, String key) {
+        AttributeValue existing = itemToTransform.get(key);
+        boolean missing = existing == null || existing.s() == null || existing.s().isEmpty();
+        if (missing) {
```
The change was applied.

RanVaknin commented:
Hi @anasatirbasa ,

I read through the PR, I have a few concerns I wanted to raise with the entire team before providing feedback. Will update you soon.

RanVaknin reviewed (changes requested):
Please refer to my comments on the PR.

anasatirbasa committed:
Added support for DynamoDbAutoGeneratedKey annotation

RanVaknin commented:
Hi @anasatirbasa,

Thanks for the follow up. I'll review and get back to you asap. 

Also a request for the future; please do not squash the commits, it makes it difficult to review the particular changes requested :) 

Thanks 🙏 
Ran~

anasatirbasa committed:
Merge branch 'master' into feature/define-dynamo-db-autogenerated-key-annotation

anasatirbasa committed:
Added support for DynamoDbAutoGeneratedKey annotation

anasatirbasa committed:
Merge branch 'master' into feature/define-dynamo-db-autogenerated-key-annotation

anasatirbasa committed:
Added support for DynamoDbAutoGeneratedKey annotation

anasatirbasa committed:
Merge branch 'master' into feature/define-dynamo-db-autogenerated-key-annotation

RanVaknin reviewed (approved):
approved on my side, needs to be reviewed and approved by someone from the DDB team.

anasatirbasa committed:
Merge branch 'master' into feature/define-dynamo-db-autogenerated-key-annotation

anasatirbasa committed:
Merge branch 'aws:master' into feature/define-dynamo-db-autogenerated-key-annotation

anasatirbasa committed:
Increased unit and integration test coverage to 100%

anasatirbasa committed:
Merge branch 'master' into feature/define-dynamo-db-autogenerated-key-annotation

anasatirbasa committed:
Increased code coverage

anasatirbasa committed:
Merge branch 'master' into feature/define-dynamo-db-autogenerated-key-annotation

amzn-erdemkemer commented on services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/extensions/AutoGeneratedKeyExtension.java:
```diff
@@ -0,0 +1,219 @@
+                                "Attribute '" + attribute + "' cannot have both @DynamoDbAutoGeneratedKey and "
+                                + "@DynamoDbAutoGeneratedUuid annotations. These annotations have conflicting behaviors "
+                                + "and cannot be used together on the same attribute.");
+                        });
```
Similar to the comment below, the conflict detection with @DynamoDbAutoGeneratedUuid currently runs on every write operation. The cost is potentially low, so not as urgent as the one below but still can be improved - which may create diff when working with large sets.

something like

```java
private final Set<TableMetadata> validatedSchemas = 
    Collections.newSetFromMap(new ConcurrentHashMap<>());

@Override
public WriteModification beforeWrite(DynamoDbExtensionContext.BeforeWrite context) {
    Collection<String> taggedAttributes = context.tableMetadata()
                                                 .customMetadataObject(CUSTOM_METADATA_KEY, Collection.class)
                                                 .orElse(null);
    if (taggedAttributes == null) {
        return WriteModification.builder().build();
    }

    // Only validate once per schema
    if (validatedSchemas.add(context.tableMetadata())) {
        validateConflicts(context, taggedAttributes);
    }
    
    // ... rest of logic
}

private void validateConflicts(DynamoDbExtensionContext.BeforeWrite context, 
                               Collection<String> taggedAttributes) {
    Collection<String> uuidTaggedAttributes = context.tableMetadata()
                                                     .customMetadataObject(UUID_EXTENSION_METADATA_KEY, Collection.class)
                                                     .orElse(Collections.emptyList());
    taggedAttributes.stream()
                    .filter(uuidTaggedAttributes::contains)
                    .findFirst()
                    .ifPresent(attribute -> {
                        throw new IllegalArgumentException(...);
                    });
}
```

anasatirbasa committed:
Tests refactoring

anasatirbasa committed:
Tests refactoring

shetsa-amzn commented on services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/extensions/AutoGeneratedKeyExtension.java:
```diff
@@ -0,0 +1,219 @@
+        }
+
+        @Override
+        public Consumer<StaticTableMetadata.Builder> modifyMetadata(String attributeName,
```
Each call passes Collections.singleton(attributeName). The addCustomMetadataObject(String key, Collection) method on StaticTableMetadata.Builder uses a merge strategy for collections. However, the existing AutoGeneratedUuidExtension uses the same pattern, so this likely works because addCustomMetadataObject with a Collection merges into the existing set under that key.

That said, this needs verification. If addCustomMetadataObject replaces rather than merges, then only the last annotated attribute would be tracked. If you have @DynamoDbAutoGeneratedKey on both PK and a GSI key, only one would get auto-generated. The existing AutoGeneratedUuidExtension would have the same latent bug, but it's typically used on a single attribute, so it's never surfaced. Worth confirming the merge behavior in StaticTableMetadata.Builder explicitly.

shetsa-amzn reviewed (commented):
The linked issue #5497 is about ConditionalCheckFailedException when using @DynamoDbAutoGeneratedUuid on a partition key combined with @DynamoDbVersionAttribute. The root cause is that AutoGeneratedUuidExtension regenerates the UUID on every write (including updates), which changes the PK and causes the version condition check to fail against a different item.

This new @DynamoDbAutoGeneratedKey extension does solve the symptom by only generating when null/empty, but the PR description should be clearer that this is a workaround/alternative rather than a fix to AutoGeneratedUuidExtension itself. Users reading #5497 might expect the existing annotation to be fixed.

shetsa-amzn reviewed (commented):
The name @DynamoDbAutoGeneratedKey implies it generates "keys" generically, but it only generates UUIDs. The V1 annotation had the same name, so there's a migration parity argument. However, in V2 we already have @DynamoDbAutoGeneratedUuid which does the same thing (generate UUIDs) but with different semantics (always vs. only-when-missing).

The real behavioral difference is "generate only when absent" vs "always regenerate." A name like @DynamoDbAutoGeneratedKeyUuid or documenting this distinction more prominently would reduce confusion. Users will inevitably ask: "which UUID annotation should I use?"

shetsa-amzn reviewed (commented):
Do we have test scenarios for the batch and transaction operations are particularly important because the extension's beforeWrite() hook behavior with batch operations could have subtle differences?

shetsa-amzn reviewed (commented):
The existing AutoGeneratedUuidExtension exposes a static create() factory method. This PR only provides builder().build(). Should keep this parity?

shetsa-amzn reviewed (commented):
Key placement validation should ideally happen during modifyMetadata() or at schema construction time, not deferred to runtime writes. I understand the StaticAttributeTag interface doesn't expose table metadata at that point, but this is worth calling out as a design limitation

anasatirbasa committed:
Addressed PR feedback and refactored tests

anasatirbasa committed:
Addressed PR feedback

anasatirbasa commented on services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/extensions/AutoGeneratedKeyExtension.java:
```diff
@@ -0,0 +1,219 @@
+                                "@DynamoDbAutoGeneratedKey can only be applied to key attributes: "
+                                + "primary partition key, primary sort key, or GSI/LSI partition/sort keys."
+                                + "Invalid placement on attribute: " + attr);
+                        });
```
Hello @amzn-erdemkemer,
Thank you for your feedback.

We cannot perform this validation inside AutoGeneratedKeyAttribute.modifyMetadata() because, at that point, the schema is still being constructed and the final TableMetadata is not yet available. The DynamoDbExtensionContext is only accessible later, during beforeWrite().

modifyMetadata() only receives the attribute name and type. At that stage, the primary key and all GSI/LSI key definitions have not yet been resolved and cannot validate whether the annotated attribute is a key.

Instead, based on your proposal, the implementation was optimized to:
-   Validate once per TableMetadata instance
-   Cache the computed set of allowed key attributes per schema

See the comment added below for for additional context: https://github.com/aws/aws-sdk-java-v2/pull/6373#discussion_r2803624818

anasatirbasa committed:
Merge branch 'master' into feature/define-dynamo-db-autogenerated-key-annotation

anasatirbasa committed:
Added tests with composite gsi

anasatirbasa committed:
Added tests with composite gsi

anasatirbasa committed:
Added tests with composite gsi

anasatirbasa committed:
Added tests with composite gsi

anasatirbasa committed:
Merge branch 'master' into feature/define-dynamo-db-autogenerated-key-annotation

amzn-erdemkemer commented on services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/enhanced/dynamodb/functionaltests/extensions/AutoGeneratedUuidExtensionTest.java:
```diff
@@ -0,0 +1,323 @@
+        List<RecordWithAutogeneratedUuid> results = mappedTable.scan().items().stream().collect(Collectors.toList());
+
+        assertThat(results.size()).isEqualTo(2);
+        isValidUuid(results.get(0).getId());
```
these are not assertions, let's make a refacotring here (applies for the rest of the file)

anasatirbasa committed:
Fixed assertions on generated uuids

anasatirbasa committed:
Merge remote-tracking branch 'origin/feature/define-dynamo-db-autogenerated-key-annotation' into feature/define-dynamo-db-autogenerated-key-annotation

anasatirbasa committed:
Merge branch 'master' into feature/define-dynamo-db-autogenerated-key-annotation

LeeroyHannigan reviewed (changes requested):
I do not see any tests for multi-attribute keys for secondary indexes? Have you tested this works, can you add a functional test case.

anasatirbasa committed:
Merge branch 'master' into feature/define-dynamo-db-autogenerated-key-annotation

anasatirbasa committed:
Merge branch 'master' into feature/define-dynamo-db-autogenerated-key-annotation

anasatirbasa committed:
Merge remote-tracking branch 'origin/feature/define-dynamo-db-autogenerated-key-annotation' into feature/define-dynamo-db-autogenerated-key-annotation

anasatirbasa committed:
Merge branch 'master' into feature/define-dynamo-db-autogenerated-key-annotation

anasatirbasa committed:
Merge branch 'master' into feature/define-dynamo-db-autogenerated-key-annotation

anasatirbasa committed:
Add support for strategy-based UUID auto-generation (@DynamoDbAutoGeneratedUuid)

RanVaknin commented on services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/extensions/annotations/DynamoDbAutoGenerateStrategy.java:
```diff
@@ -0,0 +1,35 @@
+import software.amazon.awssdk.annotations.SdkPublicApi;
+
+/**
+ * Strategy used by {@link DynamoDbAutoGeneratedUuid} to decide when UUID values are generated.
```
In the v1 SDK this strategy enum is shared between other annotations like `@DynamoDBAutoGeneratedTimestamp` that do not represent a UUID. Can we make the javadoc here generic?

RanVaknin commented on services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/extensions/AutoGeneratedUuidExtension.java:
```diff
@@ -101,42 +106,69 @@ public static AutoGeneratedUuidExtension create() {
     }
 
+    private boolean isMissingValue(AttributeValue currentValue) {
+        return currentValue == null || Boolean.TRUE.equals(currentValue.nul()) || "".equals(currentValue.s());
```
v1 did not treat an empty string as a "missing value". Whether its the right behavior is arguable. I think it's better if we stick to the existing v1 behavior of treating an empty string as a not missing.

RanVaknin reviewed (changes requested):
Thanks for all the work you have done 👍  

I created a surface API review doc and reviewed it with the Java team. It's looking good. 
I just left a few small comments on things needed fixing. 

Additionally, can we add the following test coverage?

1. Test CREATE on a non key attribute. Every CREATE test uses a key. Non keys flow through a different path (update expressions, not the key map). Please add a functional test with a plain non key String attribute annotated strategy = CREATE: generated when absent on put, preserved when present on update, generated when absent on update.
2. Test `updateItem` with `ignoreNulls(true)` + CREATE. With `ignoreNulls(true)` and a null CREATE field, the attribute is omitted from the map, so CREATE should regenerate (potentially overwriting a stored value). Can you add a functional test checking the behavior? If its a silent ovewrite we should add javadoc to reflect this edge case. 
3. Test the DynamoDB NUL missing value case. isMissingValue treats null, NUL, and "" as missing. We cover absent case and empty string case (which needs to be removed), but we dont cover the NUL case.


Lastly, the current SDK build fails because of a Japicmp error:  
```
[ERROR] Failed to execute goal com.github.siom79.japicmp:japicmp-maven-plugin:0.15.6:cmp (default) on project dynamodb-enhanced: There is at least one incompatibility: software.amazon.awssdk.enhanced.dynamodb.extensions.annotations.DynamoDbAutoGeneratedUuid.strategy():METHOD_ABSTRACT_ADDED_TO_CLASS
```

The newly added strategy was flagged by japicmp as a byte code change, but because it has a default value it should be safe. Can you please just add an exclusion to services-custom/dynamodb-enhanced/pom.xml:

```xml
  <plugin>
      <groupId>com.github.siom79.japicmp</groupId>
      <artifactId>japicmp-maven-plugin</artifactId>
      <configuration>
          <parameter>
              <excludes>
                  <exclude>software.amazon.awssdk.enhanced.dynamodb.extensions.annotations.DynamoDbAutoGeneratedUuid#strategy()</exclude>
              </excludes>
          </parameter>
      </configuration>
  </plugin>
```

anasatirbasa commented:
Hello @RanVaknin,

Thanks for the review. I've addressed the comments:

- CREATE now treats a value as missing only if it is not in the write map or it is DynamoDB NULL. An empty string is kept as-is (same as V1)
- documented that updateItem() with ignoreNulls(true) and a null CREATE field will generate a new UUID and overwrite what is already stored
- added a japicmp exclude for DynamoDbAutoGeneratedUuid#strategy()
- strategy Javadoc is generic (not UUID-only)
- on put overwrite, CREATE keeps the UUID you set, ALWAYS still generates a new one.

I also added few more scenarios:
- put with no partition key (CREATE generates it)
- LSI CREATE (generate when missing, keep when present)
- mixed-schema batch get, batch write, and transact get
- transact write with put and update in the same request

RanVaknin commented on services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/extensions/annotations/DynamoDbAutoGeneratedUuid.java:
```diff
@@ -29,7 +29,12 @@
  * {@code @DynamoDbAutoGeneratedUuid} usage.
- * Use {@link DynamoDbAutoGenerateStrategy#CREATE} when you want to generate only if the value is missing.
+ * Use {@link DynamoDbAutoGenerateStrategy#CREATE} when you want to generate only if the value is missing
+ * (absent from the write item map or DynamoDB {@code NULL}). An empty string is treated as present and is preserved.
```
nit:  This adds unnecassary info about the implementation. The following is sufficient:

```java
* Use {@link DynamoDbAutoGenerateStrategy#CREATE} when you want to generate only if the value is missing
* (absent from the write item map or DynamoDB {@code NULL}).
```

andreas-grafenberger replied:
Right, I updated the Javadoc as you proposed (commit [#9fe582f](https://github.com/aws/aws-sdk-java-v2/pull/6373/changes/9fe582fc87d905e9fd897b0d3fe30acf855e66fd)) 
cc // @RanVaknin

RanVaknin commented on services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/extensions/annotations/DynamoDbAutoGenerateStrategy.java:
```diff
@@ -18,18 +18,21 @@
 /**
- * Strategy used by {@link DynamoDbAutoGeneratedUuid} to decide when UUID values are generated.
+ * Strategy used to decide when auto-generated attribute values are produced.
+ * <p>
```
"attribute values" here is misleading, because its colliding with `AttributeValue` which is a real enhanced client container class.

The rest is an embellishment with implementation details that we don't need. 

proposed:

```java
/**
* Strategy used to decide when a new value is generated for an annotated attribute
*/
```

andreas-grafenberger replied:
Updated the Javadoc (same commit as above: [#9fe582f](https://github.com/aws/aws-sdk-java-v2/pull/6373/changes/9fe582fc87d905e9fd897b0d3fe30acf855e66fd)) 
cc // @RanVaknin

RanVaknin commented on services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/extensions/annotations/DynamoDbAutoGenerateStrategy.java:
```diff
@@ -18,18 +18,21 @@
-     * Generate a UUID only when the value is missing.
-     * Missing means the value is absent, DynamoDB {@code NULL}, or an empty string.
+     * Generate a value only when the current value is missing.
+     * Missing means the value is absent from the write item map or is DynamoDB {@code NULL}.
```
We should remove line 35 This explanation couples the enum to how the autogenerated UUID extension works. We might want to redefine what "missing" means when we implement other extensions that use this strategy enum.

andreas-grafenberger replied:
Thanks @RanVaknin, updated this Javadoc as well (same commit as above: [#9fe582f](https://github.com/aws/aws-sdk-java-v2/pull/6373/changes/9fe582fc87d905e9fd897b0d3fe30acf855e66fd))

RanVaknin commented on services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/extensions/AutoGeneratedUuidExtension.java:
```diff
@@ -101,42 +111,69 @@ public static AutoGeneratedUuidExtension create() {
+        customMetadataObject.forEach((key, strategy) -> insertUuidInItemToTransform(itemToTransform, key, strategy));
         return WriteModification.builder()
                                 .transformedItem(Collections.unmodifiableMap(itemToTransform))
                                 .build();
```
```suggestion
        Map<String, AttributeValue> itemToTransform = new HashMap<>(context.items());
        Collection<String> primaryKeys = context.tableMetadata().primaryKeys();
        boolean isUpdate = context.operationName() == OperationName.UPDATE_ITEM;
        List<UpdateAction> createOnUpdateActions = new ArrayList<>();

        customMetadataObject.forEach((key, strategy) -> {
            if (strategy == DynamoDbAutoGenerateStrategy.CREATE && isUpdate && !primaryKeys.contains(key)
                && isMissingValue(itemToTransform.get(key))) {
                itemToTransform.remove(key);
                AttributeValue seed = AttributeValue.builder().s(UUID.randomUUID().toString()).build();
                createOnUpdateActions.add(createOnlySetAction(key, seed));
            } else {
                insertUuidInItemToTransform(itemToTransform, key, strategy);
            }
        });

        WriteModification.Builder writeModification =
            WriteModification.builder().transformedItem(Collections.unmodifiableMap(itemToTransform));
        if (!createOnUpdateActions.isEmpty()) {
            writeModification.updateExpression(UpdateExpression.builder().actions(createOnUpdateActions).build());
        }
        return writeModification.build();
```

RanVaknin replied:
please dont apply suggestion through the github UI. This is meant to replace 124-128

RanVaknin commented:
Hi @anasatirbasa  and @andreas-grafenberger  thanks for the quick follow ups! 

Everything looks good, there is one last thing I want to flag before we merge, and its what I mentioned [here](https://github.com/aws/aws-sdk-java-v2/pull/6373#pullrequestreview-4954564894): 

>Test updateItem with ignoreNulls(true) + CREATE. With ignoreNulls(true) and a null CREATE field, the attribute is omitted from the map, so CREATE should regenerate (potentially overwriting a stored value). Can you add a functional test checking the behavior? If its a silent ovewrite we should add javadoc to reflect this edge case.

After re evaluating this I really think that this is a bug and we should not document it since we have an opportunity to fix it before this feature ships. On `updateItem` the extension only ever sees the caller's item map, never the row stored in DynamoDB. So `isMissingValue()` is answering "did the caller resend this field?" when the CREATE contract is really asking "does a value already exist?"  

Those are different questions, and with ignoreNulls(true) (or any partial update) they diverge, essentially for field that's present in the table looks "missing", we generate a new UUID, and the SET silently overwrites it. I think documentation here is a bandaid and I'd rather us fix this gap by letting the upstream DDB server itself make the decision on whether this field is present or not. 

I left a suggestion, please see if you can verify the behavior via tests (and change the existing tests + javadoc to match the behavior) 

After this we will be ready to merge 😄 , Thanks again for all of your work!
Ran~

anasatirbasa commented:
Hello @RanVaknin,

Thank you for the suggestion. I had a few days off, but Andreas has prepared the changes locally. We’ll go through them and make sure the implementation and tests are complete.
I’ll get back to you once they are ready for review by the end of this week. Thanks again!

anasatirbasa commented:
Hello @RanVaknin,

The proposed changes were implemented in commit: [#65a14d0](https://github.com/aws/aws-sdk-java-v2/pull/6373/changes/65a14d000eb0e1f388c3066a040b776d71e8de22). Thank you.
